### ISSUE TITLE
[Pocketbook] RFC: Input driver rewrite for direct polling

### ISSUE THREAD

**Issue by ezdiy** [koreader/koreader#6545]
The code at https://gist.github.com/ezdiy/a1e300bc05160032ea8390781f52abdd is a proof of concept rewrite of pocketbook frontend driver.

- Uses undocumented inkview polling mode running on same thread. The input side is pure ffi, similar to sdl and android.
- ~Can precisely suspend, that is, no freezes when doing something else. Currently the suspend is set to very aggressive (2s delay) for testing. Down the line, adaptive delay can be done.~ Adaptive suspend implemented.
- Some other minor features (proper multitouch handling on newer devices etc)
- ~Still rather buggy, sometimes crashing in callbacks or entering infinite loops, usually when dealing screen lock/unlock~ It's in works for me with no issues (TM) state. Still, YMMV.
- Inkview can be called directly without data races or awkward message passing

Testing would be welcome for older devices, to figure out whether this approach is workable. The point is to get rid of InkViewMain - a massive source of headaches and C boilerplate code.

**Comment by ezdiy** [comment]
In rev4:
* The hangs/crashes should be now all fixed. Turns out this was related to the weird suspend/resume spoofing, in same vein as #6536.
* Implemented adaptive suspend. This is a heuristic to have cake (battery life), and it eat too (no lag). Every tap within a time window prolongs a window of time for which we *ban* suspends, as well as the window within which further taps are accepted. If a tap occurs beyond current window, we reset the window back to 5s (ie suspend becomes very aggressive again). The effect of all this is that when you're paging through a book, doing something in UI, you won't be annoyed by suspend as it never has a chance to kick in except between first few tap or two. However once you start reading a book, the autosuspend goes in effect again - as you're quite likely to just tap once every 30 seconds, thus saving battery life. Finally, when battery is below some threshold (current 60%), we disable this logic and make autosuspend aggressive as is the default.

Things that are some more conceptual issues are now labeled FIXME, there are direct explanations about those in the code.

**Comment by roshavagarga** [comment]
@ezdiy Will be able to test out today, you reverted the 515 frontlight change between revisions though?

**Comment by roshavagarga** [comment]
@ezdiy Tested this on a 614W, seems to work without any issues so far, minus ones that existed beforehand like being unable to exit the Statistics window, due to the device having hasDPad/hasFewKeys. On that subject, I'd welcome an explanation on how those are added, as I have a few older devices I might add definitions for.

**Comment by Frenzie** [comment]
Unless there's something different about those older devices (i.e., they have a touchscreen *and* a d-pad) it's been made implicit in the no touchscreen part.

I assume it's accurate for the existing definitions which were added organically over time, though I'm not entirely convinced abstracting that particular aspect away is a good idea from a maintenance perspective. Both for clarity (the fact that you're asking probably proves the point) and just in case the assumption breaks down.

**Comment by roshavagarga** [comment]
I think one of them had an ePencil layer, which I have no idea how the code would parse, hence my question. I also don't think it should be abstracted, especially since it could be used by people that have devices that have a faulty touchscreen, but a functioning eInk layer - the dpad/hasfewkeys would somewhat allow some PB devices to still function with a custom device.lua file.

**Comment by pazos** [comment]
> In rev4:
> 
> * The hangs/crashes should be now all fixed. Turns out this was related to the weird suspend/resume spoofing, in same vein as #6536.

Good to know. I see you did some changes in the `handleMiscEv` handler. You need to still handle EVT_BACKGROUND (to save settings and notify widgets) and EVT_HIDE (to save settings). Also EVT_FOREGROUND should notify widgets the program is resumed again. 

> * Implemented adaptive suspend. This is a heuristic to have the cake (battery life), and eat it too (less lag). Every tap within a time window prolongs a window of time for which we _ban_ suspends, as well as the window within which further taps are accepted. If a tap occurs beyond current window, we reset the window back to 5s (ie suspend becomes very aggressive again). The effect of all this is that when you're paging through a book, doing something in UI, you won't be annoyed by suspend as it never has a chance to kick in except between first few tap or two. However once you start reading a book, the autosuspend goes in effect again - as you're quite likely to just tap once every 30 seconds, thus saving battery life. Finally, when battery is below some threshold (current 60%), we disable this logic and make autosuspend aggressive as is the default.

I'm not sure how big is the cake ;). In my tests and for my own usage, I would say there no significative difference. Anyways I would like to have an option to disable suspend alltogether. For long time tasks, like downloading epubs from wikipedia, a good idea would to add a function that disables sleep (something like https://github.com/koreader/koreader/blob/master/frontend/device/generic/device.lua#L344, which is only used on android). 

> Things that are sort of more conceptual issues are now labeled FIXME, there are direct explanations about those in the code.

Again, thanks for the work. I didn't try it yet as I'm still testing the changes in the linked PRs. I will wait for your PR for review and test.

I do agree with @Frenzie and @roshavagarga with stick with current device definition, as it seems easier to understand and to modify by final users.

**Comment by ezdiy** [comment]
@roshavagarga Glad to see it working on basic, this is pretty much 95% guarantee it will run on everything newer than that.

rev5:
* Put back explicit device defs, agreed that compact code ain't necessarily always readable for everyone
* Added SaveState. Note that for me, I had to do it for EVT_BACKGROUND too to get current page saved on shutdown. OS ver dependent? TODO: Investigate EVT_EXIT, par flags to determine shutdown reason more precisely?

> I'm not sure how big is the cake ;).

It will vary by device and chipset. On my PB740-2, it's 50 vs 150 of sum read time, and about 30 vs 60 with backlit (50%) on. I figured this a while ago due to long standing pbjb bug that forbade suspends. The downgrade is significant, but not really something you'll see from synthetic tests. One has to actually use the device slowly flipping through pages for few weeks with no charger to see it.

>For long time tasks, like downloading epubs from wikipedia, a good idea would to add a function that disables sleep (something like https://github.com/koreader/koreader/blob/master/frontend/device/generic/device.lua#L344, which is only used on android).

`setIgnoreInput()` is for single foreground blocking job in wait() etc. Since we're in-thread, we never enter suspend here as our poll never gets the chance to run. True background jobs are different beast - those need to signal they want suspends disabled for the duration of operation, while koreade continues to run, merely polling the background process via pipe for status now and then. Moreover, as multiple bg jobs can run simultaneously, refcount is needed, so the api would be something akin to `banSuspend(1)` on task spawn, and `banSuspend(-1)` on task finish, driver keeping count, and allowing suspend only whenever ban count is at 0. Or sentinel tables with __gc can be used to get rid of issues with refcount leaks.

Backgroundrunner would be currently user of such an api, I have some sketched code for this, but is applicable only for full koreader patchset, not a drop-in frontend driver. Luckily, there are no significant users of backgroundrunner yet, everything out there is in-thread blocking so things should work just as they are, for as long it's not an exotic plugin.

Finally, there's luasocket, and that one has ton of users. For better or worse, pocketbook already bans suspends whenever network is on, hiding this headache. If any other platform implements aggressive suspends with network on though, we'd need to monkeypatch luasocket to do `banSuspend()` on socket object creation, drop ref in socket __gc. Nasty, but more of in-theory stuff as no such device target exists in koreader as far I'm aware.

> I will wait for your PR for review and test.

Don't, the reason why I chose to go with RFC with minimal prototype is because the changes are fairly invasive when it comes to full integration. It has to drop most of old pocketbook frontend both from koreader and koreader-base, add UI plumbing for settings, apis to signal suspend bans as above etc etc. It's an entire different bunch of chores, which is why the "core" should be known to be good enough to bother with all that, ie what needs to be tested first.

**Comment by poire-z** [comment]
> everything out there is in-thread blocking 

For info, there are a few things that are run in a sub-process (via fork) while the main process is still listening for UI events (so that these few things can be interrupted if they take too long):
- dictionary lookup
- wikipedia lookup and save as EPUB
- translator
- newdownloader
- book metadata and cover extraction when needed while browsing directories with File browser (or History, File/Path Chooser, Favorites...) via CoverBrowser plugin - and when batch indexing (from File browser "+" menu)

**Comment by ezdiy** [comment]
@poire-z Yep, everything network seems to be luasocket at the end of it (can be forced via _G.__TURBO_USE_LUASOCKET__ if need be).

Indeed the forking is nasty and all over place. Somehow I had the idea that goes all through backgroundrunner, but it's p much never the case. It's probably not worth it figuring all the places that need suspend track, nor we can rely on installed timers/uilooper as that is not uniform either. Seems like I'll have to opt for nasty hacks like monkeypatch util.runInSubProcess, and ping the discovered pids every time we're about to cancel suspend ban. Hopefuly the race window of this will small enough that UI getting stuck in suspend will be rare enough.

**Comment by poire-z** [comment]
> It's probably not worth it figuring all the places

For the cases I mentionned, the main process waits for them (in a loop, while still giving control to the UI, but every so N ms, it checks again), so it's easy to find the places where to ban suspend (increase any ref count) before launching and where to unban it (decrease) when the subprocess is collected.
Or even do that in the Trapper module that wraps some of these.

For example, for the CoverBrowser case (I have some private hacks to have some kind of autostandby on my Kobo - and I indeed also disable it when WiFi on and enable it when off), I have it at these 4 points:
```diff
--- orig_koreader/plugins/coverbrowser.koplugin/bookinfomanager.lua 2020-08-22 08:31:26.000000000 +0200
+++ koreader/plugins/coverbrowser.koplugin/bookinfomanager.lua  2020-08-22 09:32:26.000000000 +0200
function BookInfoManager:collectSubprocesses()
    -- We need to regularly watch if a sub-process has terminated by
    -- calling waitpid() so this process does not become a zombie hanging
    -- around till we exit.
[...]
         else
            self.subprocesses_collector = nil
            if self.delayed_cleanup then
                self.delayed_cleanup = false
                -- No more subprocesses = no more crengine indexing, we can remove our
                -- temporary cache directory
                self:cleanUp()
            end
+           UIManager:_startAutoStandby()
        end
    end
end


function BookInfoManager:extractInBackground(files)
    if #files == 0 then
        return
    end
+    UIManager:_stopAutoStandby()
    -- Terminate any previous extraction background task that would be still running
    self:terminateBackgroundJobs()
[...]

 -- Batch extraction
function BookInfoManager:extractBooksInDirectory(path, cover_specs)
[...]
+   UIManager:_stopAutoStandby()
    while i <= nb_files do
        local filepath = files[i]
        local filename = util.basename(filepath)

        local orig_moved_offset = info.movable:getMovedOffset()
        info:free()
        info.text = T(_("Indexing %1 / %2â~@¦\n\n%3"), i, nb_files, BD.filename(filename))
[...]
    UIManager:close(info)
    info = InfoMessage:new{text = T(_("Processed %1 / %2 books."), nb_done, nb_files) .. "\n" .. T(N_("One extracted successfully.", "%1 extracted successfully.", nb_success), nb_success)}
    UIManager:show(info)
+   UIManager:_startAutoStandby()
end
```

**Comment by pazos** [comment]
> Don't, the reason why I chose to go with RFC with minimal prototype is because the changes are fairly invasive when it comes to full integration. It has to drop most of old pocketbook frontend both from koreader and koreader-base, add UI plumbing for settings, apis to signal suspend bans as above etc etc. It's an entire different bunch of chores, which is why the "core" should be known to be good enough to bother with all that, ie what needs to be tested first.

Here are my comments to your RFC :) 

1. I wouldn't mind having to deprecate support for a *very* few legacy models.
2. I do understand your idea of showcasing your changes in a single drop-in replacement. Sadly there are a few users that can actually test the changes and no active contributor has the device.
3. I would suggest to create a PR. Not one definitive, just one similar to what you did. This makes the code reviewable. Try to keep changes to the bare minimum (ie: do not add EVT_ORIENTATION if you're not working with it, leave that for later)
4. Once a PR is in the works we can release it as a package. This has proven to be the best way of testing if something is broken. Test in real life. Sadly this is not another Qt program that can be tested and debugged using VMs.

We're looking for a pocketbook platform maintainer. You seem a very good candidate for it. As far as I am concerned please go with it :+1: 

I will be more than happy to test and review PRs here and in base. I won't plan to test it before that moment.

There are some things that annoy me the most about current behaviour: wifi behaviour (tremendously broken) and heavy ghosting on partial/UI refresh mode. I will wait for the resolution to this RFC before trying to fix these :smile:

**Comment by ezdiy** [comment]
@poire-z Alright, I'll make separate PR just to get these notifiers in. If kobo (and possibly other devices) can do same style of gpio-wakeup from suspend that would be overall win if there's a ganeral way to track supend-ability.

One thing to keep these general, where the ban/unban calls should land? frontend/device, frontend/powerd, or emit an ui event, so that whoever cares (os function driver in case of pb, koreader-driver suspend plugin in case of kobo...) can listen?

**Comment by poire-z** [comment]
My stuff is a big hack, and I actually have to revert #2933 to keep everything back in UIManager() and deal with autostandby (10s) vs autosuspend (300s), with some wakup if in autostandby for 290s and need to to into suspend.
It works on my Kobo GloHD with an IR panel, that is still responsive after I do this to be in standby:
```lua
if not self:isWifiEnabled() then
    if self.auto_suspend_sec > 0 then
        local wake_in_sec = self.auto_suspend_sec - self.auto_standby_sec
        self:mylog("  programming wakealarm for in ".. wake_in_sec)
        self:_writeToSysFile("/sys/class/rtc/rtc0/wakealarm", "0")
        self:_writeToSysFile("/sys/class/rtc/rtc0/wakealarm", "+"..wake_in_sec)
    end
    self:_writeToSysFile("/sys/power/state", "standby") -- here, we're in standby
    -- here, we're out of standby because screen touched, or wakealarm triggered
    self:_writeToSysFile("/sys/class/rtc/rtc0/wakealarm", "0")
end
self:_startAutoStandby() -- restart autostandby
```
(just sharing that bit in case someone on Kobo would want to have a go at it - but there's some whole other stuff that (re)scheduleIn an action to keep track of when to go into standby - unlike on PB I guess where it's the system that decides it?)
But I have no interest in merging that into the codebase and adapting it for after #2933 where the suspend stuff is in a plugin :/ and I don't really have the big picture in mind of what would be at play.
So, I can't really say which way to go. @NiLuJe has probably more recent experience with all that (usbms, suspend...).

My gut feeling would be that it should be all dealt with by UIManager - which is the one ensuring the main loop - and that relay events - but possibly not the one catching events.
So, callers could do UIManager:preventStandby()/allowStandby() (like in my above patch), and UIManager could then ensure its own stuff, call some device/powerd implementation, broadcast events like "BeforeStandby" so other code that would need to react can just implement :onBeforeStandby().

(But how I don't want to get involved in this ! :)

**Comment by NiLuJe** [comment]
The IR grid is only powered down by the NTX global flag (controlled via the extended-state sysfs knob) IIRC). With all the race conditions that entails (which is why we always manually wake it up on resume ;p).

I haven't checked on newer devices, but there's a fair chance that's not the case on capacitive touchscreen, which makes this fairly unusable ;).

**Comment by poire-z** [comment]
Ok, so good, no auto-standby for Kobo :) 

But how would you generally see the hooks for some kind of generic standby, whether it is ensured by some external system or ensured by some plugin or some core or device stuff - and must prevented/re-allowed at some points by koreader ?
```
UIManager:allowStandby()
UIManager:preventStandby()
```
these could call device specific implemenations in Device and/or broadcast events for plugins ensuring it - (and/or broadcast events onBeforeStandby/onAfterStandby to the other application code - but I guess the point of standby ensured by external system or even plugins is that it just happens :) there's no before action as this would delay it...).

But I think that even if the PB implementation call it "suspend", in the core code we should call it "standby" (or something else), as "suspend" has already some meaning (screensaver + flush + sleep).

**Comment by ezdiy** [comment]
@poire-z agreed on the naming, it's definitely standby behaviorally.

On PB, they call is suspend or sleep because what it does internally is just `echo mem > /proc/sys/power` done by monitor whenever input timer for foreground app expires (usually 2-5 seconds since last evinput), unless is is banned via `iv_sleepmode(0)`.

I notice that interaction between standby/suspend is messy on kobo. PB OS monitor has a nifty way of doing it, IMO, which could be worth emulating in koreader:

When you hit lock button, all it does is turn off the wake gpio pins (via various means), and notifies running app it is now in background (but you can still blit to FB if you really want), paints zzz on screen. Aside from muted input, the OS still runs as usual though.

The "real" suspend is then simply induced by the standby logic that triggers due to lack of input. This has the nifty effect that you can lock the device at any time, and it will keep doing whatever it is busy doing until it's finished, which subsequently unbans standby, and that triggeres "no input, and no background stuff to do condition", and goes to true sleep.

As for API, I'll just stick it into UIManager:prevent/allowStandby that bumps/pulls the refcount, emitting an event whenever it goes from or to 0, as listeners are interested in edge of the "can i try sleep at this point?" condition, not individual bans.

This way anything that can implement standby logic can do so on their own - be it plugin directly juggling power states like you do on kobo, or device driver that signals the OS. @NiLuJe does that sound about right, or is there some special casing needed?

**Comment by NiLuJe** [comment]
Assuming the hardware cooperates (we've had a long and varied history of... weirdness on that front on NTX boards), that sounds sensible, yeah.

**Comment by j4w0r** [comment]
@ezdiy, thank you very much! Using koreader on PB after these changes is a blast :)
